### PR TITLE
Added Jasmine and Karma for testing

### PR DESCRIPTION
**Pull request by OisinNolan**
Wrote a sample function and unit tests to check that Jasmine and Karma were working. I've used JSDOM to make a mock DOM that can be manipulated for testing. If this is not suitable, another option may be to use the DOM provided by the browser that Karma runs the tests in, which also seems to work.

Also implemented the changes requested by Damien in the previous PR, which I thought I had pushed there but I must have forgotten.

### CODE REVIEW

File: karma.conf.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    logLevel: config.LOG_INFO,
+
+
+    // enable / disable watching file and executing tests whenever any file changes
+    autoWatch: true,
+
+
+    // start these browsers
+    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
+    browsers: ['Chrome'],
```
Comment (AlexLloyd0) [inline karma.conf.js:54]: We will want the library to work with other browsers. I'd suggest adding Firefox for now as that should be easy and add a TODO comment to test on other browsers. Edge/IE may be hard to test on your Chromebook but they're also more likely to have strange/different behaviour!

File: package.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  "main": "accessible-name.js",
+  "directories": {
+    "doc": "docs"
+  },
+  "devDependencies": {
+    "@types/jasmine": "^3.5.11",
+    "@types/jsdom": "^16.2.3",
+    "@types/node": "^14.0.14",
+    "common-js": "^0.3.8",
+    "dom-parser": "^0.1.6",
```
Comment (AlexLloyd0) [inline package.json:14]: Is this still used?

File: spec/TestSpec.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { getTextNodeAccessibleName } from '../src/accessible-name';
+import { JSDOM } from 'jsdom';
```
Comment (AlexLloyd0) [inline spec/TestSpec.ts:2]: I've done a bit more research into JSDOM and I don't think it's what we want to use for these tests. Various DOM APIs are implemented in pure JavaScript which is running in a browser in these tests. But that means that we're not actually using the browser's implementation so the test isn't really running in Chrome.

The good news is that if you remove lines 6 and 7 it _might_ just work?

File: spec/TestSpec.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { getTextNodeAccessibleName } from '../src/accessible-name';
```
Comment (AlexLloyd0) [inline spec/TestSpec.ts:1]: Please name .ts files in snake_case for better Closure compatibility http://go/tsstyle#file-names.

I also suggest having a _test.ts file for each normal file, so `accessible_name.ts` and `accessible_name_test.ts`.

File: spec/TestSpec.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+describe('The getTextNodeAccessibleName function', function() {
+
+  const dom: JSDOM = new JSDOM('<!DOCTYPE html><body></body>');
+  const document: Document = dom.window.document;
+  beforeEach(function() {
+    document.body.innerHTML = '';
+  });
+
+  it('returns text content if there is any', function() {
+    document.body.innerHTML = '<div id=\'foo\'>Hello there</div>';
```
Comment (AlexLloyd0) [inline spec/TestSpec.ts:13]:
```suggestion
    document.body.innerHTML = '<div id="foo">Hello there</div>';
```

File: spec/TestSpec.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    const textNode: Node = document.getElementById('foo')!.childNodes[0];
+    expect(getTextNodeAccessibleName(textNode)).toBe('Hello there');
+  });
+
+  it('returns empty string if there is no text content', function() {
+    document.body.innerHTML = '<div id=\'foo\'></div>';
```
Comment (AlexLloyd0) [inline spec/TestSpec.ts:19]:
```suggestion
    document.body.innerHTML = '<div id="foo"></div>';
```

File: spec/TestSpec.ts (hunk 4 above)
Comment (AlexLloyd0) [reply spec/TestSpec.ts:1]: Also http://go/tsstyle#code-location suggests putting `_test` files in the same directory as the code under test.

File: spec/support/jasmine.json [hunk 7]
```diff
@@ -0,0 +1,11 @@
+{
+  "spec_dir": "spec",
+  "spec_files": [
+    "**/*[sS]pec.js"
+  ],
+  "helpers": [
+    "helpers/**/*.js"
+  ],
```
Comment (AlexLloyd0) [inline spec/support/jasmine.json:8]:
```suggestion
```

File: src/accessible-name.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// Example function used to experiment with testing & DOM access/manipulation
```
Comment (AlexLloyd0) [inline src/accessible-name.ts:1]: Use JSDoc style here go/tsstyle#jsdoc-vs-comments

File: tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (AlexLloyd0) [inline tsconfig.json:1]: I haven't reviewed this, I assume it hasn't changed from #1?

File: tslint.json [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        true,
+        "check-branch",
+        "check-decl",
+        "check-operator",
+        "check-separator",
+        "check-type"
+      ]
+    }
+  }
+
```
Comment (AlexLloyd0) [inline tslint.json:66]:
```suggestion

```

**Comment by AlexLloyd0** [comment]
In general try to ensure you address review comments in the original PR :)

File: spec/TestSpec.ts (hunk 3 above)
Comment (OisinNolan) [reply spec/TestSpec.ts:2]: I think this did work when I tried it in the past, and it seems to use the browser DOM API through Karma. I'll double check this to make sure.

File: spec/TestSpec.ts (hunk 4 above)
Comment (OisinNolan) [reply spec/TestSpec.ts:1]: Will do!

File: src/accessible-name.ts (hunk 8 above)
Comment (OisinNolan) [reply src/accessible-name.ts:1]: Will do! Thanks

File: karma.conf.js (hunk 1 above)
Comment (AlexLloyd0) [reply karma.conf.js]: Once you've resolved a comment you please click 'Resolve conversation' so I can tell when you're done applying changes

File: tslint.json [hunk 11]
```diff
@@ -63,4 +63,3 @@
       ]
     }
   }
-
```
Comment (AlexLloyd0) [inline tslint.json]: +newline.

Is Visual Studio stripping these lines when it formats files?

**Pull request merged**